eventhubs/receive_eph.go

```diff
@@ -0,0 +1,114 @@
+package eventhubs
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"sync"
+	"time"
+
+	aad "github.com/Azure/azure-amqp-common-go/aad"
+	eventhubs "github.com/Azure/azure-event-hubs-go"
+	eph "github.com/Azure/azure-event-hubs-go/eph"
+	storageLeaser "github.com/Azure/azure-event-hubs-go/storage"
+	azure "github.com/Azure/go-autorest/autorest/azure"
+
+	// imports within this repo
+	"github.com/Azure-Samples/azure-sdk-for-go-samples/helpers"
+	storage "github.com/Azure-Samples/azure-sdk-for-go-samples/storage"
+)
+
+func ReceiveViaEPH() {
+	ctx := context.Background()
+
+	// get a struct with metadata about the Azure environment
+	azureEnv, err := azure.EnvironmentFromName("AzurePublicCloud")
+	if err != nil {
+		log.Fatalf("could not get azure.Environment struct: %s\n", err)
+	}
+
+	// create an access token provider using AAD principal defined in environment
+	tokenProvider, err := aad.NewJWTProvider(aad.JWTProviderWithEnvironmentVars())
+	if err != nil {
+		log.Fatalf("failed to configure AAD JWT provider: %s\n", err)
+	}
+
+	// create a storage account and container to maintain dictionary of leases
+	// and checkpoints
+	log.Printf("creating storage account\n")
+	_, err = storage.CreateStorageAccount(ctx, storageAccountName)
+	if err != nil {
+		log.Fatalf("could not create storage account: %s\n", err)
+	}
+	log.Printf("creating storage container\n")
+	_, err = storage.CreateContainer(ctx, storageAccountName, storageContainerName)
+	if err != nil {
+		log.Fatalf("could not create storage container: %s\n", err)
+	}
+
+	// use helper method to exchange AAD credentials for SAS token
+	cred, err := storageLeaser.NewAADSASCredential(
+		helpers.SubscriptionID(),
+		helpers.ResourceGroupName(),
+		storageAccountName,
+		storageContainerName,
+		storageLeaser.AADSASCredentialWithEnvironmentVars())
+	if err != nil {
+		log.Fatalf("could not prepare a storage credential: %s\n", err)
+	}
+
+	// create a leaser and checkpointer backed by a storage container
+	leaserCheckpointer, err := storageLeaser.NewStorageLeaserCheckpointer(
+		cred,
+		storageAccountName,
+		storageContainerName,
+		azureEnv)
+	if err != nil {
+		log.Fatalf("could not prepare a storage leaserCheckpointer: %s\n", err)
+	}
+
+	// use all of the above to create an Event Processor
+	p, err := eph.New(
+		ctx,
+		nsName,
+		hubName,
+		tokenProvider,
+		leaserCheckpointer,
+		leaserCheckpointer,
+		eph.WithNoBanner())
+	if err != nil {
+		log.Fatalf("failed to create EPH: %s\n", err)
+	}
+	defer p.Close(context.Background())
+
+	// set up a handler for the Event Processor which will receive a single
+	// message, print it to the console, and allow the process to exit
+
+	// set up wait group to wait for expected message
+	var wg sync.WaitGroup
+	wg.Add(1) // wait for one message
```

Comment:
I'd like to steer away from using a `sync.WaitGroup` here, because they error out if they get decremented more times than expected.

If it were me, I'd use channels, something like this:

``` Go
receivedEvent := make(chan struct{})

handle := func(ctx context.Context, event *eventhubs.Event) error {
    fmt.Printf("received: %s\n", string(event.Data))
    receivedEvent <- struct{}{}
    return nil
}

// stuff

<-receivedEvent
```

Both approaches are prone to leaking the resources captured via the closure of this func. However, using channels will prevent misused code from panicing. There's probably some argument to be had here, because health checks on paniced code may be easier to detect than memory leaks, but for the sake of example code I'd like them to be a little more robust to people tinkering with the number of messages sent, etc.